crates/stages/src/stages/bodies.rs

```diff
@@ -106,45 +106,22 @@ impl<DB: Database, D: BodyDownloader, C: Consensus> Stage<DB> for BodyStage<D, C
         // Get id for the first transaction in the block
         let mut first_tx_id = db.get_first_tx_id(starting_block)?;
 
-        // Cursor used to look up headers for block pre-validation
-        let mut header_cursor = db.cursor::<tables::Headers>()?;
-
         // NOTE(onbjerg): The stream needs to live here otherwise it will just create a new iterator
         // on every iteration of the while loop -_-
         let mut bodies_stream = self.downloader.bodies_stream(bodies_to_download.iter());
         let mut highest_block = previous_block;
-        while let Some((block_number, header_hash, body)) =
+        while let Some(block) =
             bodies_stream.try_next().await.map_err(|err| StageError::Internal(err.into()))?
         {
-            // Fetch the block header for pre-validation
-            let block = BlockLocked {
-                header: SealedHeader::new(
-                    header_cursor
-                        .seek_exact((block_number, header_hash).into())?
-                        .ok_or(DatabaseIntegrityError::Header {
-                            number: block_number,
-                            hash: header_hash,
-                        })?
-                        .1,
-                    header_hash,
-                ),
-                body: body.transactions,
-                ommers: body.ommers.into_iter().map(|header| header.seal()).collect(),
-            };
-
-            // Pre-validate the block and unwind if it is invalid
-            self.consensus
-                .pre_validate_block(&block)
-                .map_err(|err| StageError::Validation { block: block_number, error: err })?;
```

Comment:
This was a bug before. If a peer sent us a bad block they could force us to unwind (because the Pipeline will catch the StageError::Validation and unwind). But what we actually want is to detect this error in the downloader, discard that body, and penalize the peer (as done below).